src/base/rss/rss_feed.cpp

```diff
@@ -0,0 +1,418 @@
+    int i = -1;
+    foreach (const QJsonValue &jsonVal, jsonArr) {
+        ++i;
+        if (!jsonVal.isObject()) {
+            Logger::instance()->addMessage(
+                        QString("Couldn't load RSS article '%1#%2'. Invalid data format.").arg(m_url).arg(i)
+                        , Log::WARNING);
+            continue;
+        }
+
+        auto article = Article::fromJsonObject(this, jsonVal.toObject());
+        if (article && !addArticle(article))
+            delete article;
+    }
+}
+
+void Feed::loadArticlesLegacy()
+{
+    SettingsPtr qBTRSSFeeds = Profile::instance().applicationSettings(QStringLiteral("qBittorrent-rss-feeds"));
+    QVariantHash allOldItems = qBTRSSFeeds->value("old_items").toHash();
+
+    foreach (const QVariant &var, allOldItems.value(m_url).toList()) {
+        auto article = Article::fromVariantHash(this, var.toHash());
+        if (article && !addArticle(article))
+            delete article;
+    }
+}
+
+void Feed::store()
+{
+    QJsonArray jsonArr;
+    foreach (Article *article, m_articles)
+        jsonArr << article->toJsonObject();
+
+    m_session->dataFileStorage()->store(m_dataFileName, QJsonDocument(jsonArr).toJson());
+}
+
+bool Feed::addArticle(Article *article)
+{
+    Q_ASSERT(article);
+
+    if (m_articles.contains(article->guid()))
+        return false;
+
+    // Insertion sort
+    auto articleDateLessThan = [](Article *article, QDateTime date)
+    {
+        return article->date() < date;
+    };
+    const int maxArticles = m_session->maxArticlesPerFeed();
+    auto lowerBound = std::lower_bound(m_articlesByDate.begin(), m_articlesByDate.end()
+                                       , article->date(), articleDateLessThan);
+    if ((lowerBound - m_articlesByDate.begin()) >= maxArticles)
+        return false; // we reach max articles
+
+    m_articles[article->guid()] = article;
+    m_articlesByDate.insert(lowerBound, article);
+    if (!article->isRead()) {
+        increaseUnreadCount();
+        connect(article, &Article::read, this, &Feed::handleArticleRead);
+    }
+    emit newArticle(article);
+
+    if (m_articlesByDate.size() > maxArticles)
+        removeOldestArticle();
+
+    return true;
+}
+
+void Feed::removeOldestArticle()
+{
+    auto oldestArticle = m_articlesByDate.takeLast();
+    m_articles.remove(oldestArticle->guid());
+    emit articleAboutToBeRemoved(oldestArticle);
+    bool isRead = oldestArticle->isRead();
+    delete oldestArticle;
+
+    if (!isRead)
+        decreaseUnreadCount();
+}
+
+void Feed::increaseUnreadCount()
+{
+    ++m_unreadCount;
+    emit unreadCountChanged(this);
+}
+
+void Feed::decreaseUnreadCount()
+{
+    Q_ASSERT(m_unreadCount > 0);
+
+    --m_unreadCount;
+    emit unreadCountChanged(this);
+}
+
+void Feed::downloadIcon()
+{
+    // Download the RSS Feed icon
+    // XXX: This works for most sites but it is not perfect
+    auto iconUrl = QString("http://%1/favicon.ico").arg(QUrl(m_url).host());
```

Comment:
This should be using the same scheme as the feed url for the icon. This is an existing bug that I had fixed in one of my PRs. https://github.com/qbittorrent/qBittorrent/pull/6368/files#diff-5890b862f225705b8036dc3fbfc71f79L321